### PR TITLE
OBGM-643 Unable to display view page of invoice created in other currency than USD

### PR DESCRIPTION
**Pull request by alannadolny**

### CODE REVIEW

**Review by kchelstowski — APPROVED** [review]

File: grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy [hunk 1]
```diff
@@ -149,7 +149,7 @@ class Invoice implements Serializable {
         BigDecimal currentExchangeRate
         String defaultCurrencyCode = Holders.config.openboxes.locale.defaultCurrencyCode
         if (currencyUom?.code != defaultCurrencyCode) {
-            currentExchangeRate = UnitOfMeasureConversion.conversionRateLookup(defaultCurrencyCode, currencyUom?.code).list()
+            currentExchangeRate = UnitOfMeasureConversion.conversionRateLookup(defaultCurrencyCode, currencyUom?.code).get()
```
Comment (awalkowiak) [inline grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy:152]: Are we sure that we always will have one active conversion rate for a specific currency (I am not sure how `uniqueResult = true` already works there)? Perhaps it is worth improving the named query to sort by last updated date?

[the remark this answers is not in the document]
Comment (alannadolny) [reply grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy:152]: As I saw we can have more than one active conversation rate at the same time (it,s a bit odd 🤔), so I added this sorting by updated date to avoid some different behavior.

[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy:152]: I agree with @awalkowiak. I think the reason why you can add multiple conversions for the same to/from uoms is because ... lazy. 

For now you can either add uniqueResult = true to the named query. Use the minTransactionDate named query on the TransactionFact domain.

https://github.com/openboxes/openboxes/blob/e19d7df63e4a4b394e7b39000fdb84bf42acfb7b/grails-app/domain/org/pih/warehouse/reporting/TransactionFact.groovy#L53

Or you could do something like this 
```
UnitOfMeasureConversion.conversionRateLookup(defaultCurrencyCode, currencyUom?.code).list()[0]
```
Or this (don't think it's necessary, but who knows.
```
UnitOfMeasureConversion.conversionRateLookup(defaultCurrencyCode, currencyUom?.code).list([max: 1])[0]
```
You'll need to test the result set count = (0, 1, N) cases to make sure this returns an empty list when there are no records and the correct one when there are more than one records.

**Review by jmiranda — CHANGES_REQUESTED** [review]

File: grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (alannadolny) [reply grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy:152]: @jmiranda this `get` works as expected.
When there are more than 2 conversion rates, the one with a newer updated date is being taken. When there is no conversion rate the null is returned and it also works as expected.
conversionRateLookup includes order by updated date descending and unique result.

As an example:
I have three conversions from USD to British Pounds
image
The latest update is this one with a 1.39 conversion rate.
So, going to invoice made in British Pounds
image
280.78 / 202 = 1.39
So that's the correct conversion rate.

[the remark this answers is not in the document]
Comment (jmiranda) [reply grails-app/domain/org/pih/warehouse/invoice/Invoice.groovy:152]: Awesome. Sorry I must have missed the uniqueResult = true earlier.

**Review by jmiranda — APPROVED** [review]
